storage: prep ChangeReplicasTrigger for atomic rebalancing

tbg opened the pull request:
The implementation plan for atomic rebalancing is to first expose a
testing-only API that lets us carry out general atomic membership changes
(i.e. not limited to a single swap which is all that we expect to want to
use in production in 19.2).

This allows us to really stress the machinery in general and also prevents
more API headaches down the road should we want to support more general
changes. Besides, the change to an API that supports a single swap only is
already similarly invasive.

This is the first in a sequence of PRs/commits that plumb the API through
the system (but without actually supporting atomic changes yet). The whole
pipeline that needs to be plumbed is

- `AdminChangeReplicasRequest` which calls into
- `(*Replica).ChangeReplicas` which calls into
- the internal code adding learners, sending them snapshots, etc, and
which finally evaluates into a
- `ChangeReplicasTrigger` that needs to eventually be translated
into a
- `raftpb.ConfChangeV2` (we use a v1 `ConfChange` today).

This particular change deprecates the `Replica` and `ChangeType` field of
the trigger, and introduces instead two slices of ReplicaDescriptors, one
for added and one for removed replicas. This will be easy to translate into
a `ConfChangeV2` and seemed more idiomatic (and less error-prone) than
including the base desc and computing a sequence of changes from that
whenever it is needed. It also matches what the higher levels of the API
will do.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/39485)
<!-- Reviewable:end -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> all nits. Thanks for the consumable change.

> Reviewed 14 of 14 files at r1, 5 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39485#-:-Lm55XtO62DfzJ-_wlhk:b-jqd68s)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @tbg)

___
*[docs/generated/settings/settings.html, line 125 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39485#-Lm52Qb29q2gi0Hf0pHc:-Lm52Qb29q2gi0Hf0pHd:b-rjag40) ([raw file](https://github.com/cockroachdb/cockroach/blob/4cd451f11dea0960e3f99430a5f5050bfebfa456/docs/generated/settings/settings.html#L125)):*
> ```HTML
> <tr><td><code>trace.lightstep.token</code></td><td>string</td><td><code></code></td><td>if set, traces go to Lightstep using this token</td></tr>
> <tr><td><code>trace.zipkin.collector</code></td><td>string</td><td><code></code></td><td>if set, traces go to the given Zipkin instance (example: '127.0.0.1:9411'); ignored if trace.lightstep.token is set</td></tr>
> <tr><td><code>version</code></td><td>custom validation</td><td><code>19.1-8</code></td><td>set the active cluster version in the format '<major>.<minor>'</td></tr>
> ```

nit: it looks like this landed in the wrong commit.
___
*[pkg/roachpb/data.go, line 1310 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39485#-Lm52t1f45o09IrWFQ8d:-Lm52t1f45o09IrWFQ8e:b-alqmdc) ([raw file](https://github.com/cockroachdb/cockroach/blob/c548ab88030a673b99860b6e57118e3fdfa8c3b8/pkg/roachpb/data.go#L1310)):*
> ```Go
> 	if crt.Desc != nil {
> 		nextReplicaID = crt.Desc.NextReplicaID
> 		// NB: we don't want to mutate InternalReplicas, so we don't call
> ```

Will this change once https://github.com/cockroachdb/cockroach/pull/39489 is merged?
___
*[pkg/roachpb/data.go, line 1327 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39485#-Lm53QXl9KsvHSLjlBOf:-Lm53QXl9KsvHSLjlBOg:b95apqp) ([raw file](https://github.com/cockroachdb/cockroach/blob/c548ab88030a673b99860b6e57118e3fdfa8c3b8/pkg/roachpb/data.go#L1327)):*
> ```Go
> 		fmt.Fprintf(&chgS, "%s%s", REMOVE_REPLICA, removed)
> 	}
> 	return fmt.Sprintf("%s: after=%s next=%d", chgS.String(), afterReplicas, nextReplicaID)
> ```

nit: you might as well keep using the `strings.Builder`:

```
fmt.Fprintf(&chgS, ": after=%s next=%d", afterReplicas, nextReplicaID)
return chgS.String()
```
___
*[pkg/roachpb/data.proto, line 165 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39485#-Lm51r_pCWUYrlnXBp1_:-Lm51r_pCWUYrlnXBp1a:b7zp1t0) ([raw file](https://github.com/cockroachdb/cockroach/blob/c548ab88030a673b99860b6e57118e3fdfa8c3b8/pkg/roachpb/data.proto#L165)):*
> ```protobuf
>   option (gogoproto.goproto_stringer) = false;
> 
>   // TODO(tbg): remove when possible. 19.2 won't actively use this any more,
> ```

Mind mentioning when this will be possible?
___
*[pkg/roachpb/data.proto, line 189 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39485#-Lm523ycD4IdqUO0XAm1:-Lm523ycD4IdqUO0XAm2:b-yd7o4t) ([raw file](https://github.com/cockroachdb/cockroach/blob/c548ab88030a673b99860b6e57118e3fdfa8c3b8/pkg/roachpb/data.proto#L189)):*
> ```protobuf
>   // set when this field is used). Must not specify the same replica more than once or
>   // overlap with the added replicas above.
>   repeated ReplicaDescriptor internal_removed_replicas = 7 [(gogoproto.nullable) = false];
> ```

The plan is to remove the "internal" prefix from these eventually, right?
___
*[pkg/settings/cluster/cockroach_versions.go, line 509 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39485#-Lm52DFmA-hpcNScEqxB:-Lm52DFmA-hpcNScEqxC:bnrknu0) ([raw file](https://github.com/cockroachdb/cockroach/blob/c548ab88030a673b99860b6e57118e3fdfa8c3b8/pkg/settings/cluster/cockroach_versions.go#L509)):*
> ```Go
> 	},
> 	{
> 		// VersionAtomicChangeReplicasTrigger is <TODO>.
> ```

Reminder to fill this in.
___
*[pkg/storage/replica_raft.go, line 293 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39485#-Lm54vbm4zNj0SgC6TD-:-Lm54vbm4zNj0SgC6TD0:b-2zhur2) ([raw file](https://github.com/cockroachdb/cockroach/blob/c548ab88030a673b99860b6e57118e3fdfa8c3b8/pkg/storage/replica_raft.go#L293)):*
> ```Go
> 		prefix = false
> 
> 		// Ensure that we aren't trying to remove ourselves from the range without
> ```

It seems strange to me that this lives here. I think I tried moving it down to batch evaluation at one point, but ran into issues because the commit triggers are a little panic-y. If you're working in this area and have it all paged in, do you mind taking a pass at giving this check a better home?


<!-- Sent from Reviewable.io -->

tbg committed:
storage: prep ChangeReplicasTrigger for atomic rebalancing

The implementation plan for atomic rebalancing is to first expose a
testing-only API that lets us carry out general atomic membership
changes (i.e. not limited to a single swap which is all that we
expect to want to use in production in 19.2).

This allows us to really stress the machinery in general and also
prevents more API headaches down the road should we want to support
more general changes. Besides, the change to an API that supports
a single swap only is already similarly invasive.

This is the first in a sequence of PRs/commits that plumb the API
through the system (but without actually supporting atomic changes yet).
The whole pipeline that needs to be plumbed is

- `AdminChangeReplicasRequest` which calls into
- `(*Replica).ChangeReplicas` which calls into
- the internal code adding learners, sending them snapshots, etc, and
  which finally evaluates into a
- `ChangeReplicasTrigger` that needs to eventually be translated
  into a
- `raftpb.ConfChangeV2` (we use a v1 `ConfChange` today).

This particular change deprecates the `Replica` and `ChangeType` field
of the trigger, and introduces instead two slices of ReplicaDescriptors,
one for added and one for removed replicas. This will be easy to
translate into a `ConfChangeV2` and seemed more idiomatic (and less
error-prone) than including the base desc and computing a sequence of
changes from that whenever it is needed. It also matches what the higher
levels of the API will do.

As a drive-by, also remove ConfChangeContext.Replica, which has been
unused essentially forever (<2.0).

Release note: None

tbg reviewed (commented):
Thank you! All addressed.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39485#-:-Lm5z7wBA3krHXRUulyY:b-z1wlab)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[pkg/roachpb/data.go, line 1310 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39485#-Lm52t1f45o09IrWFQ8d:-Lm5x6F0Bv-WPl1QvdlZ:bx25kg1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c548ab88030a673b99860b6e57118e3fdfa8c3b8/pkg/roachpb/data.go#L1310)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Will this change once https://github.com/cockroachdb/cockroach/pull/39489 is merged?
</blockquote></details>

Yes! Added a TODO to revisit.
___
*[pkg/roachpb/data.proto, line 165 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39485#-Lm51r_pCWUYrlnXBp1_:-Lm5xbsD2mOMEE3a8PCI:b-ajcwps) ([raw file](https://github.com/cockroachdb/cockroach/blob/c548ab88030a673b99860b6e57118e3fdfa8c3b8/pkg/roachpb/data.proto#L165)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Mind mentioning when this will be possible?
</blockquote></details>

Added a reference to #39182.
___
*[pkg/roachpb/data.proto, line 189 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39485#-Lm523ycD4IdqUO0XAm1:-Lm5y4cS-SYI6PrEKQXC:b-1ui7sb) ([raw file](https://github.com/cockroachdb/cockroach/blob/c548ab88030a673b99860b6e57118e3fdfa8c3b8/pkg/roachpb/data.proto#L189)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

The plan is to remove the "internal" prefix from these eventually, right?
</blockquote></details>

Yep, added TODO to the deprecated fields.
___
*[pkg/storage/replica_raft.go, line 293 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39485#-Lm54vbm4zNj0SgC6TD-:-Lm5w-3A5A9ytCYRJUMv:b3j3i89) ([raw file](https://github.com/cockroachdb/cockroach/blob/c548ab88030a673b99860b6e57118e3fdfa8c3b8/pkg/storage/replica_raft.go#L293)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

It seems strange to me that this lives here. I think I tried moving it down to batch evaluation at one point, but ran into issues because the commit triggers are a little panic-y. If you're working in this area and have it all paged in, do you mind taking a pass at giving this check a better home?
</blockquote></details>

I share the sentiment, will take another look when I'm less pressed for time. This *should* be straightforward.


<!-- Sent from Reviewable.io -->

tbg commented:
bors r=nvanbenschoten
